pkg/cmd/server/middleware.go

```diff
@@ -0,0 +1,207 @@
+package server
+
+import (
+	"fmt"
+
+	"github.com/authzed/spicedb/pkg/util"
+
+	"google.golang.org/grpc"
+)
+
+// MiddlewareChain describes an ordered sequence of middlewares that can be modified
+// with one or more MiddlewareModification. This struct is used to facilitate the
+// creation and modification of gRPC middleware chains
+type MiddlewareChain struct {
+	chain []ReferenceableMiddleware
+}
+
+// NewMiddlewareChain creates a new middleware chain given a zero or more named middlewares.
+// An error will be returned in case validation of the NamedMiddlewares fail.
+func NewMiddlewareChain(mw ...ReferenceableMiddleware) (MiddlewareChain, error) {
+	if err := validate(mw); err != nil {
+		return MiddlewareChain{}, err
+	}
+	return MiddlewareChain{chain: mw}, nil
+}
+
+// MiddlewareModification describes an operation to modify a MiddlewareChain
+type MiddlewareModification struct {
+	// Dependency is used to define with respect to which middleware an operation is performed
```

Comment:
Maybe we should name it `DependencyMiddlewareName` to be explicit? Not sure its necessary

Reply:
Should also comment this should be empty for replace all